### PR TITLE
Port to TPM2.0 support for the current tpm functions

### PR DESCRIPTION
**Pull request by leonjia0112**
This PR is for porting TPM interface from TPM1.2 to TPM2.0 is issue #44.
Notes for this PR
* create_deep_quote is working in progress will be added in the future PR, since vTPM is currently not available.
* Current supporting are the functions which are needed for TPM initialization of the key lime agent
* Change testing suite to TPM 2.0 functions testing
* Added additional tpm2.0 function which is needed in the restful request handling

### CODE REVIEW

**Comment by frozencemetery** [comment]
This looks good so far.  As discussed in person, please update this PR once you have `create_deep_quote` working.

File: src/tpm.rs [hunk 1]
```diff
@@ -196,144 +206,82 @@ Following are function from tpm_quote.py program
...
-        quote_path,
-    );
-
-    // RUN
-    let (_, quote_raw) = run(command, Some(quote_path))?;
-    let mut quote_return = String::from("d");
-    quote_return.push_str(&quote_raw);
-    Ok(quote_return)
+    Err(KeylimeTpmError::new_tpm_rust_error(
+        "Deep quote in progress.",
```
Comment (frozencemetery) [inline src/tpm.rs]: This looks good so far.  As discussed in person, please update this PR once you have `create_deep_quote` working.

Comment (leonjia0112) [reply src/tpm.rs]: Will you merge this one first or until all the TPM2.0 functions are implemented?

**Comment by lukehinds** [comment]
Is this replacing 1.2 with 2.0?

I am not strongly attached to using 1.2 as its an old spec, but the python-keylime project supports both 1.2 and 2.0 - they make a query here:

https://github.com/keylime/python-keylime/blob/d277757199a555f78dde69cca366557e232f9cfa/keylime/tpm_obj.py#L49-L68

We should perhaps consider mirroring them and keeping your existing 1.2 work, and just creating a tpm2.rs built around a conditional similar to the above.

**Comment by frozencemetery** [comment]
I'm opposed to writing new code that uses 1.2.  MIT is welcome to keep supporting 1.2 for their users, but it cannot provide the security guarantees we want out of keylime (due to known-weak cryptography).

**Comment by lukehinds** [comment]
@frozencemetery fair enough points, so as said not strongly minded here. TPM 1.2 chips are becoming legacy now and will become few and far between each day. Let's start as we mean to go on then and just use 2.0 then.

**Comment by leonjia0112** [comment]
Initial review requested
@frozencemetery @lukehinds @jetwhiz @mbestavros 
Changed implementation of read_tpmdata() function, pass is now passed.

**Comment by leonjia0112** [comment]
> Please don't request review unless CI is passing, thanks!

It passed yesterday \(T.T)/.

**Comment by leonjia0112** [comment]
Test passed. Ready for a review.
@lukehinds @frozencemetery @mbestavros 
Thanks!

File: src/tpm.rs [hunk 2]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+ *  Write data to a Non-Volatile (NV) index.
+ */
+fn tpm_write_key_nvram(key: String) -> Result<(), KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let mut tf = NamedTempFile::new()?;
+    tf.write_all(key.as_bytes())?;
+    let attrs = "ownerread|policywrite|ownerwrite";
+    let tf_path = temp_file_get_path(&tf)?;
+
+    run(format!("tpm2_nvdefine -x 0x1500018 -a 0x40000001 -s {} -t \"{}\" -p {} -P {}", common::BOOTSTRAP_KEY_SIZE, attrs, owner_pw, owner_pw), None)?;
```
Comment (lukehinds) [inline src/tpm.rs]: We might need to implement a check for legacy_tools, if we want to work with rhel 7:

https://github.com/keylime/keylime/blob/master/keylime/tpm2.py#L1031

https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L49-L53

You can see where this is a decision based on the

File: src/tpm.rs [hunk 3]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+) -> Result<(), KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let aik = get_tpm_metadata_content("aik")?;
+    let aik_name = get_tpm_metadata_content("aik_name")?;
+
+    if !aik.is_empty() && !aik_name.is_empty() {
+        let aik_handle = get_tpm_metadata_content("aik_handle")?;
+
+        let (ret_out, _) =
+            run("tpm2_getcap -c handles-persisitent".to_string(), None)?;
```
Comment (lukehinds) [inline src/tpm.rs]: This is incorrectly spelt: `handles-persisitent` will fail, should be:

`run("tpm2_getcap -c handles-persistent".to_string(), None)?;`

File: src/tpm.rs [hunk 4]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+        let ret_out_map: Value = serde_yaml::from_str(&ret_out)?;
+        if ret_out_map[&aik_handle].is_null() {
+            run(
+                format!(
+                    "tpm2_evictcontrol -a o -c {} -P {}",
```
Comment (lukehinds) [inline src/tpm.rs]: Going to need to check the tools version as before:

https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L345-L348

File: src/tpm.rs [hunk 5]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+        ));
+    }
+    let aik_pw = random_password(20)?;
+
+    let tf = NamedTempFile::new()?;
+    let aik_pub_path = temp_file_get_path(&tf)?;
+    let paths = vec![aik_pub_path];
+    let (ret_out, f_out) = run(
+        format!(
+            "tpm2_createak -C {} -k - -G {} -D {} -s {} -p {} -f pem -e {} -P {} -o {}",
```
Comment (lukehinds) [inline src/tpm.rs]: another legacy tools check need (I know its a PIA, but we will need this for RHEL 7 support);

https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L551-L554

File: src/tpm.rs [hunk 6]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+        set_tpm_metadata_content("owner_pw", &owner_pw);
+    }
+
+    let ek_pw = random_password(20)?;
+    let tf = NamedTempFile::new()?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let paths: Vec<&str> = vec![tf_path];
+    let (ret_out, f_out) = run(
+        format!(
+            "tpm2_createek -c -G {} -p {} -P {} -o {} -e {}",
```
Comment (lukehinds) [inline src/tpm.rs]: https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L373-L376

File: src/tpm.rs [hunk 7]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+            ek_asym_alg, tf_path, ek_pw, owner_pw, owner_pw
+        ),
+        Some(paths),
+    )?;
+
+    let ek_tpm = f_out.get(tf_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "Createek output to file content is missing",
```
Comment (lukehinds) [inline src/tpm.rs]: Could we maybe include `tf_path` in the error message to help the user debug? I would rephrase as well:

`Createek file output to file $tf_path cannot be found`

File: src/tpm.rs [hunk 8]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+ */
+fn tpm_get_pub_ek() -> Result<(), KeylimeTpmError> {
+    let ek_handle = get_tpm_metadata_content("ek_handle")?;
+    let tf = NamedTempFile::new()?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let paths = vec![tf_path];
+
+    let (_, ret_file_map) = run(
+        format!(
+            "tpm2_readpublic -c {} -o {} -f pem",
```
Comment (lukehinds) [inline src/tpm.rs]: legacy tools check

https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L439-L442

File: src/tpm.rs [hunk 9]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+            hex::encode(ek_handle),
+            tf_path
+        ),
+        Some(paths),
+    )?;
+
+    let ek = ret_file_map.get(tf_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "tpm2_readpublic fail, ek key is missing is output file.",
```
Comment (lukehinds) [inline src/tpm.rs]: grammar needs fixing up:

`"tpm2_readpublic failed, output file &tf_path cannot be found and in turn ek key loaded.",`

File: src/tpm.rs [hunk 10]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+ *
+ *  Retrieve all the persistent object and remove those object.
+ */
+fn flush_keys() -> Result<(), KeylimeTpmError> {
+    let (ret_out, _) = run("tpm2_getcap -c handles-persistent".into(), None)?;
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let ret_out_val: Value = serde_yaml::from_str(&ret_out)?;
+    let ret_out_map: &Map<String, Value> =
+        ret_out_val.as_object().ok_or_else(|| {
+            KeylimeTpmError::new_tpm_rust_error("Output invalid.")
```
Comment (lukehinds) [inline src/tpm.rs]: could do with a more descriptive error message here, the user won't know what Output is invalid.

File: src/tpm.rs [hunk 11]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+        })?;
+    ret_out_map.iter().for_each(|(k, _)| {
+        run(
+            format!(
+                "tpm2_evictcontrol -a o -c {} -P {}",
```
Comment (lukehinds) [inline src/tpm.rs]: you guessed it :)
https://github.com/keylime/keylime/blob/master/keylime/tpm2.py#L515

File: src/tpm.rs [hunk 12]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+        "sha1" => 160,
+        "sha256" => 256,
+        "sha384" => 384,
+        "sha512" => 512,
+        _ => 0,
+    };
+
+    let value_num: i64 =
+        ret_out_map[hash_alg_str][pcrval].as_i64().ok_or_else(|| {
+            KeylimeTpmError::new_tpm_rust_error("Invalid pcr valur.")
```
Comment (lukehinds) [inline src/tpm.rs]: spelling mistake `valur`

File: src/tpm.rs [hunk 13]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+    let aik_pw = get_tpm_metadata_content("aik_hanlde")?;
+    let mut tf = NamedTempFile::new()?;
+    tf.write_all(keyblob.as_bytes())?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let secdir = secure_mount::mount()?;
+    let tf_sec = NamedTempFile::new_in(secdir)?;
+    let tf_sec_path = temp_file_get_path(&tf_sec)?;
+    let (ret_out, f_out) = run(
+        format!(
+            "tpm2_activatecredential -c {} -C {} -f {} -o {} -P {} -E {}",
```
Comment (lukehinds) [inline src/tpm.rs]: https://github.com/keylime/keylime/blob/master/keylime/tpm2.py#L709

File: src/tpm.rs [hunk 14]
```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
...
+            tf_sec_path,
+            aik_pw,
+            owner_pw
+        ),
+        Some(vec![tf_sec_path]),
+    )?;
+
+    let file_out = f_out.get(tf_sec_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "tpm_activate_identity failed with invalid file output.",
```
Comment (lukehinds) [inline src/tpm.rs]: You have the function name here, whereas I think you mean the command `tpm2_activatecredential`?

File: src/tpm.rs [hunk 15]
```diff
@@ -135,219 +615,215 @@ fn write_tpm_data(data: Value) -> Result<(), KeylimeTpmError> {
...
-        let token: Vec<&str> = line_tmp.split_whitespace().collect();
-        if token.len() == 3 {
-            if token[0] == "VendorID" && token[1] == ":" {
-                return Ok(token[2].to_string());
-            }
+pub fn is_software_tpm() -> bool {
+    match tpm_get_manufacturer() {
+        Ok(data) => data == "SW",
+        Err(e) => {
+            warn!("Fail to get tpm manufacturer information. Error {}.", e);
```
Comment (lukehinds) [inline src/tpm.rs]: *Failed to get...

File: src/tpm.rs [hunk 16]
```diff
@@ -135,219 +615,215 @@ fn write_tpm_data(data: Value) -> Result<(), KeylimeTpmError> {
...
-    );
+    let quote_path = temp_file_get_path(&quote_tf)?;
+    let sign_path = temp_file_get_path(&sign_tf)?;
+    let pcr_path = temp_file_get_path(&pcr_tf)?;
+
+    let paths = vec![quote_path, sign_path, pcr_path];
+
+    let (_, quotes) = run(
+        format!(
+            "tpm2_deluxequote -C {} -L {}:{} -q {} -m {} -s {} -p {} -G {} -P {}",
```
Comment (lukehinds) [inline src/tpm.rs]: https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L851-L854

File: src/tpm.rs [hunk 17]
```diff
@@ -135,219 +615,215 @@ fn write_tpm_data(data: Value) -> Result<(), KeylimeTpmError> {
...
-            format!("0x{}", (vpcrmask_int + (1 << common::TPM_DATA_PCR)));
-        let mut command = format!("pcrreset -ix {}", common::TPM_DATA_PCR);
-
-        //RUN
-        run(command, None)?;
-        let mut hasher = sha::Sha1::new();
-        hasher.update(data.as_bytes());
-        let data_sha1_hash = hasher.finish();
+    Err(KeylimeTpmError::new_tpm_rust_error(
+        "Deep quote in progress.",
```
Comment (lukehinds) [inline src/tpm.rs]: Is "in progress" an error?

**Comment by lukehinds** [comment]
Can we close this one now @leonjia0112 ?

**Comment by leonjia0112** [comment]
> Can we close this one now @leonjia0112 ?

sure!

**Comment by leonjia0112** [comment]
Newer implementation in #75
